lte/cloud/go/sms_ll/sms_rp.go

```diff
@@ -0,0 +1,365 @@
+		TPDU: data,
+	}, nil
+
+}
+
+func (rpue RPUserElement) MarshalBinary(msgType byte) ([]byte, error) {
+	rpu_len := len(rpue.TPDU) + 1
+	b := make([]byte, 0, rpu_len)
+	if msgType == RP_ACK || msgType == RP_ERROR { // these start with IEI
+		b = append(b, rpue.IEI)
+	}
+	b = append(b, rpue.Length)
+	b = append(b, rpue.TPDU...)
+	return b, nil
+}
+
+func (rpue *RPUserElement) UnmarshalBinary(msgType byte, input []byte) (int, error) {
+	idx := 0
+	if msgType == RP_ACK || msgType == RP_ERROR { // these start with IEI
+		rpue.IEI = input[idx]
+		idx++
+	}
+	rpue.Length = input[idx]
+	idx++
+	end := idx + int(rpue.Length)
+	rpue.TPDU = input[idx:end]
+	return end, nil
+}
+
+// RP-Cause element (TS 24.011 8.2.5.4)
+type RPCauseElement struct {
+	IEI        byte // Never serialized
+	Length     byte
+	Cause      byte
+	Diagnostic byte // Optional
+}
+
+func (rpce RPCauseElement) MarshalBinary() []byte {
+	b := []byte{rpce.Length, rpce.Cause}
+	if int(rpce.Length) == 2 {
+		b = append(b, rpce.Diagnostic)
+	}
+	return b
+}
+
+func (rpce *RPCauseElement) UnmarshalBinary(input []byte) int {
+	rpce.Length = input[0]
+	rpce.Cause = input[1]
+	if int(rpce.Length) == 2 {
+		rpce.Diagnostic = input[2]
+		return 3
+	}
+	return 2
+}
+
+type RPMessage struct {
+	MTI       byte
+	Reference byte
+
+	// Mandantory. If UE->Network, must be length 0
+	OriginatorAddress RPAddressElement
+
+	// Mandantory. If Network->UE, must be length 0
+	DestinationAddress RPAddressElement
+
+	// Mandantory for RP-DATA, includes TPDU. If RP-ACK or RP-ERROR, must
+	// include IEI.
+	UserData RPUserElement
+
+	// Mandantory for RP-ERROR
+	Cause RPCauseElement
+}
+
+func (rpm RPMessage) MarshalBinary() ([]byte, error) {
+	b := []byte{rpm.MTI, rpm.Reference}
+
+	rpmt, err := rpm.MsgType()
+	if err != nil {
+		return nil, err
+	}
+
+	switch rpmt {
+	case RP_DATA:
+		oa_data, err := rpm.OriginatorAddress.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		da_data, err := rpm.DestinationAddress.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		b = append(b, oa_data...)
+		b = append(b, da_data...)
+
+		ud_data, _ := rpm.UserData.MarshalBinary(RP_DATA)
+		b = append(b, ud_data...)
+	case RP_ERROR:
+		cause_data := rpm.Cause.MarshalBinary()
+		b = append(b, cause_data...)
+	case RP_ACK:
+		// Do nothing
+	default:
+		return nil, SMSRPError("Invalid RP Message Type")
+	}
+
+	return b, nil
+
+}
+
+func (rpm *RPMessage) UnmarshalBinary(input []byte) error {
+	if len(input) < 2 {
+		return SMSRPError("SMS-RP Message too short")
+	}
+
+	idx := 0
+	rpm.MTI = input[idx]
+	idx++ // 1
+	rpm.Reference = input[idx]
+	idx++ // 2
+
+	rpmt, err := rpm.MsgType()
+	if err != nil {
+		return err
+	}
+
+	switch rpmt {
+	case RP_DATA:
+		// The next two IEs should be adddresses in this case. So, get the lengths and pass to unmarshal
+		rpm.OriginatorAddress = *new(RPAddressElement)
+		rpm.DestinationAddress = *new(RPAddressElement)
+		rpm.UserData = *new(RPUserElement)
```

Comment:
you don't actually have to initialize these fields. in Go the zero type for a struct is the struct with all its fields set to their zero values: https://medium.com/rungo/structures-in-go-76377cc106a2